packages/ember-data/lib/system/snapshot.js

```diff
@@ -0,0 +1,364 @@
+    var callDeprecate = true;
+
+    Ember.defineProperty(this, 'constructor', {
+      get: function() {
+        // Ugly hack since accessing error.stack (done in `Ember.deprecate()`)
+        // causes the internals of Chrome to access the constructor, which then
+        // causes an infinite loop if accessed and calls `Ember.deprecate()`
+        // again.
+        if (callDeprecate) {
+          callDeprecate = false;
+          Ember.deprecate('Usage of `snapshot.constructor` is deprecated, use `snapshot.type` instead.');
+          callDeprecate = true;
+        }
+
+        return this.type;
+      }
+    });
+  } else {
+    this.constructor = this.type;
+  }
+}
+
+Snapshot.prototype = {
+  constructor: Snapshot,
+
+  /**
+    The underlying record for this snapshot. Can be used to access methods and
+    properties defined on the record.
+
+    Example
+
+    ```javascript
+    var json = snapshot.record.toJSON();
+    ```
+
+    @property record
+    @type {DS.Model}
+  */
+  record: null,
+
+  /**
+    The type of the underlying record for this snapshot, as a subclass of DS.Model.
+
+    @property type
+    @type {subclass of DS.Model}
+  */
+  type: null,
+
+  /**
+    The name of the type of the underlying record for this snapshot, as a string.
+
+    @property typeKey
+    @type {String}
+  */
+  typeKey: null,
+
+  /**
+    Returns the value of an attribute.
+
+    Example
+
+    ```javascript
+    var post = store.createRecord('post', { author: 'Tomster', title: 'Ember.js rocks' });
+    var snapshot = post._createSnapshot();
+
+    snapshot.attr('author'); // => 'Tomster'
+    snapshot.attr('title'); // => 'Ember.js rocks'
+    ```
+
+    Note: Values are loaded eagerly and cached when the snapshot is created.
+
+    @method attr
+    @param {String} keyName
+    @return {Object} The attribute value or undefined
+  */
+  attr: function(keyName) {
+    if (keyName in this._attributes) {
+      return this._attributes[keyName];
+    }
+    throw new Ember.Error("Model '" + Ember.inspect(this.record) + "' has no attribute named '" + keyName + "' defined.");
+  },
+
+  /**
+    Returns all attributes and their corresponding values.
+
+    Example
+
+    ```javascript
+    var post = store.createRecord('post', { author: 'Tomster', title: 'Ember.js rocks' });
+    var snapshot = post._createSnapshot();
+
+    snapshot.attributes(); // => { author: 'Tomster', title: 'Ember.js rocks' }
+    ```
+
+    @method attributes
+    @return {Array} All attributes for the current snapshot
+  */
+  attributes: function() {
+    return Ember.copy(this._attributes);
+  },
+
+  /**
+    Returns the id of the snapshot's record
+
+    Example
+
+    ```javascript
+    var post = store.push('post', { id: 1, author: 'Tomster', title: 'Ember.js rocks' });
+    var snapshot = post._createSnapshot();
+
+    snapshot.id(); // => '1'
+    ```
+
+    @method id
+    @return {String} The ID of the snapshot's record
+  */
+  id: function() {
```

Comment:
I'm nervous because now if you do `record.id` it will return the id of the record. This could break serializers that are relying on this behavior.